### ISSUE TITLE
How to build ResNet network to train with cuda

### ISSUE THREAD

**Issue by HangJie720** [haanjack/mnist-cudnn#4]
Although deep learning frameworks(e.g., TF, Pytorch,...) are commonly used to train models, I still want to try to build a Resnet model using Cudnn/ Cublas library for the deep learning training.
Could you give me some suggestions about how to create classic deep learning network(e.g., ResNet, ...) for cuda.
Thank you very much!
Actually, for resnet18, only Pad/AddV2/Reshape/Mean Ops are need to implement.

**Comment by haanjack** [comment]
Your friend are [CUDNN API](https://docs.nvidia.com/deeplearning/sdk/cudnn-api/index.html), [CUBLAS API](https://docs.nvidia.com/cuda/cublas/index.html), and frameworks you are familiar with. I recommend you to read Caffe or Caffe2 blobs implemented codes how they have built the frameworks from CUDA and these APIs.

Basically, you can implement your own CUDNN/CUBLAS network by replacing TF/PyTorch network. 

For examples, there are Conv, BatchNorm2d, and ReLU are used in PyTorch as shown below.
```python
class BasicBlock(nn.Module):
    expansion = 1

    def __init__(self, inplanes, planes, stride=1, downsample=None, groups=1,
                 base_width=64, dilation=1, norm_layer=None):
        super(BasicBlock, self).__init__()
        if norm_layer is None:
            norm_layer = nn.BatchNorm2d
        if groups != 1 or base_width != 64:
            raise ValueError('BasicBlock only supports groups=1 and base_width=64')
        if dilation > 1:
            raise NotImplementedError("Dilation > 1 not supported in BasicBlock")
        # Both self.conv1 and self.downsample layers downsample the input when stride != 1
        self.conv1 = conv3x3(inplanes, planes, stride)
        self.bn1 = norm_layer(planes)
        self.relu = nn.ReLU(inplace=True)
        self.conv2 = conv3x3(planes, planes)
        self.bn2 = norm_layer(planes)
        self.downsample = downsample
        self.stride = stride

    def forward(self, x):
        identity = x

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)

        if self.downsample is not None:
            identity = self.downsample(x)

        out += identity
        out = self.relu(out)

        return out
```
And when you trace down PyTorch operations, you can reach to how PyTorch used CUDNN/CUBLAS APIs for the operations. Then, what you can do is to use that into your implementation without other PyTorch's interfacing codes. Luckily, CUDNN provides such operations like TF/PyTorch does. But you should write forward/backwards parts for the block.

For example, we can use 
`[cudnnBatchNormalizationForwardTraininig()](https://docs.nvidia.com/deeplearning/sdk/cudnn-api/index.html#cudnnBatchNormalizationForwardTraining)`, `[cudnnDivisiveNormalizationBackward()](https://docs.nvidia.com/deeplearning/sdk/cudnn-api/index.html#cudnnDivisiveNormalizationBackward)`, and `[cudnnBatchNormalizationForwardInference()](https://docs.nvidia.com/deeplearning/sdk/cudnn-api/index.html#cudnnBatchNormalizationForwardInference)` for BatchNorm layer implementation.

If you consider to use my codes, you can add `BatchNorm2d` with forward and backward. Then you can write residual blocks from the layers already implemented as you does with PyTorch/TF.

Or you can write a sequence of CUDNN/CUBLAS API callings without such operation hierarchy.

**Comment by HangJie720** [comment]
Thank you very much, now I try to reference TF kernels implementation to perform FusedBatchNormalization and other operators.
I have a question, if the model structure contains branches, as show as follows,
![](https://tva1.sinaimg.cn/large/007S8ZIlgy1ghli8ceqs9j30cr0jewhf.jpg)
![](https://tva1.sinaimg.cn/large/007S8ZIlgy1ghliasmt3qj30fu0ki3zt.jpg)
how the backward is calculated？gradients backward from two convolution branches is to do the add operation first? And then to execute Relu/Identity backward?

**Comment by HangJie720** [comment]
@haanjack, I want to know how to adjust parameters due to misconvergence when training resnet18. Could you help me check out what the problem is ? I have email to you code and model.

**Comment by haanjack** [comment]
@HangJie720 
Could you share your code after fork and share me the branch?
I saw your email a few days ago, but I cannot find it from mailbox.. I'm sorry about that.

BTW, how did you soleved the backward propagation problem?
I think that would be simply element-wise some as you suggested in the second picture.

Currently, I'm working on to integrate newer CUDNN APIs into this repository which merges [CONV+ACT](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnConvolutionBiasActivationForward). Because the book publisher want to update with the newer libraries.

**Comment by HangJie720** [comment]
@haanjack , my code is share to https://github.com/HangJie720/DeepLearning-Training-Cuda
I found it feasible to train Resnet using the Adam optimizer, and the backward propagation flow is consistent with what I envisioned by looking at the source code of TensorFlow(https://github.com/tensorflow/tensorflow/blob/master/tensorflow/cc/framework/gradients.cc).
Q1: You can try to train vgg16(https://github.com/HangJie720/DeepLearning-Training-Cuda/tree/master/mnist-cudnn-vgg16) , which is not convergence.
Q2: The other question that struck me as odd was, If the first layer is pad implemented by eigen, a simple network(2 layers of conv+bn+relu+pool) is also not convergence, you can try to run https://github.com/HangJie720/DeepLearning-Training-Cuda/blob/master/mnist-cudnn-simple/train.cpp by setting the first layer is pad. I think the following two approaches are equivalent, however the first case doesn't converge, the second case is ok.
```
model.add_layer(new Pad("pad1", {0, 0, 0, 0, 2, 2, 2, 2}, 0)); //[1,1,28,28] -> [1,1,32,32]
model.add_layer(new Conv2D("conv1", 32, 5, 1, 0)); //[1,1,32,32] -> [1,32,28,28]
```
```
model.add_layer(new Conv2D("conv1", 32, 5, 1, 2)); //[1,1,28,28] -> [1,32,28,28]
```

**Comment by HangJie720** [comment]
For Pad， I try to implement pad op forward/backward with cuda reference to [caffe](https://github.com/17cspence/caffe/blob/pad_layer/src/caffe/layers/pad_layer.cu) source code,  model can be convergence. I don't know why pad implemented by eigen appears to misconvergence. 
Thank you very much for taking the time to help me checkout If pad gradient backward propagation code exists error.

However, pad self-implemented by cuda is time comsumping(with kernel config is <<<16, 512>>>), profile results are shown as follows,
![](https://tva1.sinaimg.cn/large/007S8ZIlgy1gi5q7tjwwoj31020g5tgb.jpg)

I want to know how is the pad in convolution implemented efficiently? Could you give me some suggestions?

**Comment by HangJie720** [comment]
The problem that the pad implemented by eigen has not converged is that the third input parameter Options_ of Eigen::Tensor supports RowMajor, ColMajor and two forms. The default is ColMajor, but the problem is solved after modifying it to RowMajor.
But for eigen::pad，it's performance is not effective to pad implemented by cuda kernel.
